### ISSUE TITLE
Creating audio stack just to handle OPTIONS message

### ISSUE THREAD

**Issue by mikhail-barg** [alfredh/baresip#265]
Here's a call stack for handling `OPTIONS` message:
```
ourlib.dll!fd_listen(int fd, int flags, void(*)(int, void *) fh, void * arg) Line 584	C
ourlib.dll!udp_thread_attach(udp_sock * us) Line 696	C
ourlib.dll!udp_listen(udp_sock * * usp, const sa * local, void(*)(const sa *, mbuf *, void *) rh, void * arg) Line 376	C
ourlib.dll!udp_range_listen(rtp_sock * rs, const sa * ip, unsigned short min_port, unsigned short max_port) Line 241	C
ourlib.dll!rtp_listen(rtp_sock * * rsp, int proto, const sa * ip, unsigned short min_port, unsigned short max_port, char enable_rtcp, void(*)(const sa *, const rtp_header *, mbuf *, void *) recvh, void(*)(const sa *, rtcp_msg *, void *) rtcph, void * arg) Line 330	C
ourlib.dll!stream_sock_alloc(stream * s, int af) Line 254	C
ourlib.dll!stream_alloc(stream * * sp, const config_avt * cfg, call * call, sdp_session * sdp_sess, const char * name, int label, const mnat * mnat, mnat_sess * mnat_sess, const menc * menc, menc_sess * menc_sess, const char * cname, void(*)(const rtp_header *, mbuf *, void *) rtph, void(*)(rtcp_msg *, void *) rtcph, void * arg) Line 302	C
ourlib.dll!audio_alloc(audio * * ap, const config * cfg, call * call, sdp_session * sdp_sess, int label, const mnat * mnat, mnat_sess * mnat_sess, const menc * menc, menc_sess * menc_sess, unsigned int ptime, const list * aucodecl, void(*)(int, char, void *) eventh, void(*)(int, const char *, void *) errh, void * arg) Line 703	C
ourlib.dll!call_alloc(call * * callp, const config * cfg, list * lst, const char * local_name, const char * local_uri, account * acc, ua * ua, const call_prm * prm, const sip_msg * msg, call * xcall, dnsc * dnsc, void(*)(call *, call_event, const char *, void *) eh, void * arg) Line 593	C
ourlib.dll!ua_call_alloc(call * * callp, ua * ua, vidmode vidmode, const sip_msg * msg, call * xcall, const char * local_uri) Line 447	C
ourlib.dll!handle_options(ua * ua, const sip_msg * msg) Line 473	C
ourlib.dll!request_handler(const sip_msg * msg, void * arg) Line 553	C
ourlib.dll!sip_recv(sip * sip, const sip_msg * msg) Line 266	C
ourlib.dll!udp_recv_handler(const sa * src, mbuf * mb, void * arg) Line 339	C
ourlib.dll!udp_read(udp_sock * us, int fd) Line 219	C
ourlib.dll!udp_read_handler(int flags, void * arg) Line 232	C
ourlib.dll!fd_handler(re * re, int fd, int flags) Line 240	C
ourlib.dll!fd_poll(re * re) Line 843	C
ourlib.dll!re_main(void(*)(int) signalh) Line 979	C
ourlib.dll!ReMainThreadFunc(void * lpThreadParameter) Line 237	C
```
See that `handle_options()` calls `ua_call_alloc()` which in turn calls `audio_alloc()` to handle RTP sessions which seem to be completely redundant in case of `OPTIONS` message.

(Initially reported in https://github.com/creytiv/re/issues/61)

**Comment by alfredh** [comment]
hi @mikhail-barg many thanks for this report.

a question to you and everybody else; do you think that the OPTIONS response
should contain SDP or not? Removing the whole SDP is the easiest solution.

from RFC 3261, 11.2 Processing of OPTIONS Request

```
   A message body MAY be sent, the type of which is determined by the
   Accept header field in the OPTIONS request (application/sdp is the
   default if the Accept header field is not present).  If the types
   include one that can describe media capabilities, the UAS SHOULD
   include a body in the response for that purpose.  Details on the
   construction of such a body in the case of application/sdp are
   described in [13].
```

I think it is common for most SIP useragent implementations to NOT include the SDP.
For example SEMS:

```
$ sipsak -v -s sip:music@iptel.org

checking message for completness...
message is complete
SIP/2.0 200 OK
Via: SIP/2.0/TCP 127.0.1.1:59699;branch=z9hG4bK.32343619;rport=34435;alias;received=91.64.95.178
From: sip:sipsak@127.0.1.1:59699;tag=3504dbe0
To: sip:music@iptel.org;tag=3299371A-593301EB0009AC4A-4CC47700
Call-ID: 889510880@127.0.1.1
CSeq: 1 OPTIONS
Server: Sip Express Media Server (1.1.1-aefe2 (x86_64/linux))
X-Call-ID: 729FD813-593301EB0009AF02-4CC47700
Contact: <sip:3299371A-593301EB0009AC4A-4CC47700@212.79.111.155;transport=tcp>
Content-Length: 0

```

**Comment by GGGO** [comment]
It's true that SIP OPTIONS is more use like a qualify but the RFC3261 says too:

```
11 Querying for Capabilities

   The SIP method OPTIONS allows a UA to query another UA or a proxy
   server as to its capabilities.  This allows a client to discover
   information about the supported methods, content types, extensions,
   codecs, etc. without "ringing" the other party.  For example, before
   a client inserts a Require header field into an INVITE listing an
   option that it is not certain the destination UAS supports, the
   client can query the destination UAS with an OPTIONS to see if this
   option is returned in a Supported header field.  All UAs MUST support
   the OPTIONS method.
```

I would say that baresip is doing right to reply with the sdp body because it follows the RFC; and maybe someone else is already using the querying capabilities.

If I understand correctly, this sdp allocation will never be used by the remote, so if the remote decides to finally send a INVITE, it will make a new fresh offer/answer SDP, making the SDP's OPTIONS useless.

Do we need a new parameter in the account file to control the behavior of SIP OPTIONS ?

**Comment by czarkoff** [comment]
IMO baresip is doing it right, and I don't see an issue in probing for audio and video support upon receiving OPTIONS.

**Comment by alfredh** [comment]
after some more reading I think baresip is doing the right thing. I made a small commit
that checks the Accept header. If the Accept header is set to e.g. "text/plain" the SDP
will not be included in the response.

the OPTIONS request is supposed to be handled as if there is an incoming call,
that is why we are creating a full call-state object with audio stack etc.

@mikhail-barg are you able to control the OPTIONS request in your setup?
if so you can try to set the Accept to something else that application/sdp.


/Alfred

**Comment by oej** [comment]
I think that the response should include SDP but you should not initiate the audio stack and open RTP ports. The answer to OPTIONS should be the same as for an INVITE, that's correct. But it should not affect media state. Right now we get a lot of error messages in syslog for in-dialog OPTIONs which is not good. And opening RTP ports outside of a an active dialog can cause harm...

**Comment by alfredh** [comment]
thank you @oej Olle for your view on this.

I have made a commit that implements OPTIONS handling as suggested by Olle.

could you all please test this commit and see if things are working
as expected:

https://github.com/alfredh/baresip/commit/ac510eac6d39c7d3244d6c8b08bc5fb22a147b6b


/Alfred

**Comment by alfredh** [comment]
Hi @mikhail-barg could you please verify my fix ?

**Comment by mikhail-barg** [comment]
@alfredh I'll try to merge and see how things are working, but it may take a couple of days, for I have a bit involved setup here. I'll get beck to you as soon as I get results. Thanks for looking into this!

**Comment by alfredh** [comment]
released as part of Baresip 0.5.4

I now consider this issue fixed, so I will close the ticket.
Verification is still appreciated.


/Alfred
